### PR TITLE
chore(expression): be compatible with version v2 bloom filter index

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

No need to reload the data anymore.

Closes #issue

### CODE REVIEW

File: src/query/expression/src/converts/to.rs [hunk 1]
```diff
@@ -37,3 +40,38 @@ pub fn to_schema(schema: &TableSchema) -> common_datavalues::DataSchema {
         .collect();
     common_datavalues::DataSchema::new_from(fields, schema.meta().clone())
 }
+
+pub fn scalar_to_datavalue(scalar: &Scalar) -> DataValue {
```
Comment (sundy-li) [inline src/query/expression/src/converts/to.rs:44]: @drmingdrmer  this function is re add  because we use DataValue's hash to apply old bloom filter.

**Review by drmingdrmer — APPROVED** [review]

File: src/query/storages/index/src/bloom.rs [hunk 2]
```diff
@@ -248,7 +254,15 @@ impl BlockFilter {
         };
 
         let (filter, _size) = Xor8Filter::from_bytes(filter_bytes)?;
-        if filter.contains(&target) {
+
+        let contains = if self.version == 2 {
```
Comment (BohuTANG) [inline src/query/storages/index/src/bloom.rs:258]: can we change the constant `2` to an enum value BlockBloomFilterIndexVersion::V2?

Comment (BohuTANG) [reply src/query/storages/index/src/bloom.rs:258]: This way, it will be easier for us to consider it when we refactor it in the future.

Comment (sundy-li) [reply src/query/storages/index/src/bloom.rs:258]: ok

**Review by drmingdrmer — APPROVED** [review]

File: src/query/storages/index/tests/it/filters/bloom_filter.rs [hunk 3]
```diff
@@ -63,7 +63,7 @@ fn test_bloom_filter() -> Result<()> {
     ];
     let chunks_ref = chunks.iter().collect::<Vec<_>>();
 
-    let index = BlockFilter::try_create(FunctionContext::default(), schema, &chunks_ref)?;
+    let index = BlockFilter::try_create(FunctionContext::default(), schema, 3, &chunks_ref)?;
```
Comment (BohuTANG) [inline src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: 3 --> `V3BloomBlock::VERSION`?

Comment (BohuTANG) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: I guess this test is expected to test the latest version, not `3`?

Comment (sundy-li) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: Yes, 3 is the latest bloom version.

Comment (sundy-li) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: I'll update it, there are two struct named BlockFilter

Comment (BohuTANG) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: So, when we upgrade our version number next time, this test may be missed. Can we define a latest version or another way to archive it?

Comment (Xuanwo) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs:66]: I prefer to add a `LastestBlockBloomFilterIndexVersion = BlockBloomFilterIndexVersion::V3`?

**Comment by BohuTANG** [comment]
@mergify update

File: src/query/storages/index/tests/it/filters/bloom_filter.rs (hunk 3 above)
Comment (sundy-li) [reply src/query/storages/index/tests/it/filters/bloom_filter.rs]: BlockBloomFilterIndexVersion::V3 is a enum, we need u64 here.

File: src/query/storages/table-meta/src/meta/versions.rs [hunk 4]
```diff
@@ -163,9 +163,13 @@ mod converters {
         type Error = ErrorCode;
         fn try_from(value: u64) -> Result<Self, Self::Error> {
             match value {
-                1 | 2 => Err(ErrorCode::DeprecatedIndexFormat(
+                1 => Err(ErrorCode::DeprecatedIndexFormat(
```
Comment (BohuTANG) [inline src/query/storages/table-meta/src/meta/versions.rs:166]: Looks better to define constant instead the int `1` `2` `3`.

Comment (sundy-li) [reply src/query/storages/table-meta/src/meta/versions.rs:166]: this is already the lowest level now and it's the same logic as before.

We can define constant field name  `const VersionV1 = 1`, but it's too redundant

**Comment by BohuTANG** [comment]
Test this PR, old bloom filter(V2) also works as new(V3):
```
mysql> SELECT UserID FROM hits_expr WHERE UserID = 435090932899640449; --new
+--------------------+
| userid             |
+--------------------+
| 435090932899640449 |
| 435090932899640449 |
| 435090932899640449 |
| 435090932899640449 |
+--------------------+
4 rows in set (0.10 sec)
Read 779848 rows, 5.95 MiB in 0.074 sec., 10.51 million rows/sec., 80.20 MiB/sec.

mysql> SELECT UserID FROM hits  WHERE UserID = 435090932899640449; -- old
+--------------------+
| userid             |
+--------------------+
| 435090932899640449 |
| 435090932899640449 |
| 435090932899640449 |
| 435090932899640449 |
+--------------------+
4 rows in set (0.13 sec)
Read 496176 rows, 3.79 MiB in 0.117 sec., 4.24 million rows/sec., 32.34 MiB/sec.
```

**Pull request merged**